### PR TITLE
feat: [CDS-74848]: Added the DTOs and entity class for Global template

### PR DESCRIPTION
**Pull request by shivamnegi94**
I have split the PR into small PRs.
Adding the dtos, entity class and mapper in this PR.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51363)

### CODE REVIEW

**Comment by shivamnegi94** [comment]
trigger uts

File: 878-ng-common-utilities/src/main/java/io/harness/gitaware/helper/GitAwareEntityHelper.java [hunk 1]
```diff
@@ -349,4 +351,46 @@ public void validateRepo(
...
+    String commitId =
+        isNullOrDefault(gitContextRequestParams.getCommitId()) ? "" : gitContextRequestParams.getCommitId();
+
+    GitAwareContextHelper.setIsDefaultBranchInGitEntityInfoWithParameter(branch);
+
+    String filePath = gitContextRequestParams.getFilePath();
+    if (isNullOrDefault(filePath)) {
+      throw new InvalidRequestException("No file path provided.");
+    }
+    validateReamMeFilePathHasCorrectExtension(filePath);
```
Comment (utkarshchoubey-harness) [inline 878-ng-common-utilities/src/main/java/io/harness/gitaware/helper/GitAwareEntityHelper.java:370]: typo -
validateReadMeFileHasCorrectExtension

File: template-service/service/src/main/java/io/harness/template/mappers/NGTemplateDtoMapper.java [hunk 2]
```diff
@@ -438,4 +440,47 @@ public boolean parseLoadFromCacheHeaderParam(String loadFromCache) {
...
+
+  public EntityGitDetails getEntityGitDetails(GlobalTemplateEntity globalTemplateEntity) {
+    return globalTemplateEntity.getStoreType() == null
+        ? EntityGitDetailsMapper.mapEntityGitDetails(globalTemplateEntity)
+        : globalTemplateEntity.getStoreType() == StoreType.REMOTE
+        ? GitAwareContextHelper.getEntityGitDetailsFromScmGitMetadata()
+        : EntityGitDetails.builder().build();
+  }
+
+  public CacheResponseMetadataDTO getCacheResponse(GlobalTemplateEntity templateEntity) {
```
Comment (utkarshchoubey-harness) [inline template-service/service/src/main/java/io/harness/template/mappers/NGTemplateDtoMapper.java:480]: Are you adding Caching support as well here ?

File: template-service/service/src/main/java/io/harness/template/services/NGTemplateSchemaServiceImpl.java [hunk 3]
```diff
@@ -300,4 +302,56 @@ private JsonNode getStaticYamlSchemaFromResource(String accountIdentifier, Strin
     }
     return getTemplateSchema(accountIdentifier, projectIdentifier, orgIdentifier, scope, templateChildType, entityType);
   }
+
+  public void validateYamlSchemaInternal(GlobalTemplateEntity globalTemplateEntity) {
```
Comment (utkarshchoubey-harness) [inline template-service/service/src/main/java/io/harness/template/services/NGTemplateSchemaServiceImpl.java:306]: this is the exact same function , we can reuse the old function? 
Or create a overloaded 
or maybe put the common logic in diff function and use it across

File: template-service/service/src/main/java/io/harness/template/events/GlobalTemplateUpdateEvent.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import java.util.HashMap;
+import java.util.Map;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@OwnedBy(CDC)
+@Getter
+@NoArgsConstructor
+public class GlobalTemplateUpdateEvent implements Event {
```
Comment (utkarshchoubey-harness) [inline template-service/service/src/main/java/io/harness/template/events/GlobalTemplateUpdateEvent.java:33]: can we utilise the same Events we have for templates ?
As nearly the code is entirely duplicated 
Also does it makes sense to use same events ?
cc - @archit-harness @sribalij

**Review by utkarshchoubey-harness — CHANGES_REQUESTED** [review]

File: template-service/service/src/main/java/io/harness/template/events/GlobalTemplateUpdateEvent.java (hunk 4 above)
Comment (shivamnegi94) [reply template-service/service/src/main/java/io/harness/template/events/GlobalTemplateUpdateEvent.java:33]: We should not use same event.

File: template-service/service/src/main/java/io/harness/template/mappers/NGTemplateDtoMapper.java (hunk 2 above)
Comment (shivamnegi94) [reply template-service/service/src/main/java/io/harness/template/mappers/NGTemplateDtoMapper.java:480]: I am creating same support as of template have right now.

File: template-service/service/src/main/java/io/harness/template/services/NGTemplateSchemaServiceImpl.java (hunk 3 above)
Comment (shivamnegi94) [reply template-service/service/src/main/java/io/harness/template/services/NGTemplateSchemaServiceImpl.java:306]: Entity type is different so i have overriden the method

**Comment by shivamnegi94** [comment]
trigger uts

**Review by utkarshchoubey-harness — APPROVED** [review]

**Comment by shivamnegi94** [comment]
trigger uts

**Review by archit-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51363)* status: 10 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`template-service/service/src/main/java/io/harness/template/events/GlobalTemplateUpdateEvent.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/51363#gh-1284214238:-NbNo6438bhckl0MJi9B:b-lushab) ([raw file](https://github.com/harness/harness-core/blob/eda180668ca5c788dfeeb442c852249c1adc503e/template-service/service/src/main/java/io/harness/template/events/GlobalTemplateUpdateEvent.java#L33)):*
<details><summary><i>Previously, shivamnegi94 (Shivam Negi) wrote…</i></summary><blockquote>

We should not use same event.
</blockquote></details>

yeah having separate event is better if record is different, if fields are same, one thing you can do is put inside abstract class, that will remove duplicate code

**Review by archit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51363)* status: 10 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`template-service/service/src/main/java/io/harness/template/entity/GlobalTemplateEntity.java` line 127 at r3](https://reviewable.io/reviews/harness/harness-core/51363#-NbNoSqs-9rq2U6Ex59F:-NbNoSqs-9rq2U6Ex59G:bb4pi2) ([raw file](https://github.com/harness/harness-core/blob/bf5d9a89054770c3d7d70236898c483a5b0e3da6/template-service/service/src/main/java/io/harness/template/entity/GlobalTemplateEntity.java#L127)):*
> ```Java
> 
>   public static List<MongoIndex> mongoIndexes() {
>     return ImmutableList.<MongoIndex>builder()
> ```

why so many indexes are required?

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 8 files at r1, 3 of 4 files at r3, 2 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51363)* status: 10 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`template-service/service/src/main/java/io/harness/template/entity/GlobalTemplateEntity.java` line 127 at r3](https://reviewable.io/reviews/harness/harness-core/51363#-NbNoSqs-9rq2U6Ex59F:-NbOxRy16W7Dup4IsbPc:bysfpnd) ([raw file](https://github.com/harness/harness-core/blob/bf5d9a89054770c3d7d70236898c483a5b0e3da6/template-service/service/src/main/java/io/harness/template/entity/GlobalTemplateEntity.java#L127)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why so many indexes are required?
</blockquote></details>

how both index can be unique?
___
*[`template-service/service/src/main/java/io/harness/template/services/NGTemplateSchemaServiceImpl.java` line 306 at r1](https://reviewable.io/reviews/harness/harness-core/51363#gh-1284211814:-NbOxXOH70y-m2uN1TPj:b-cphpbj) ([raw file](https://github.com/harness/harness-core/blob/eda180668ca5c788dfeeb442c852249c1adc503e/template-service/service/src/main/java/io/harness/template/services/NGTemplateSchemaServiceImpl.java#L306)):*
<details><summary><i>Previously, shivamnegi94 (Shivam Negi) wrote…</i></summary><blockquote>

Entity type is different so i have overriden the method
</blockquote></details>

@utkarshchoubey-harness can you validate existing schema is not impacted?

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 4 files at r3, 1 of 3 files at r5, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51363)* status: 10 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by utkarshchoubey-harness — APPROVED** [review]

**Comment by shivamnegi94** [comment]
trigger uts

**Comment by shivamnegi94** [comment]
trigger review

**Pull request merged**